SoftBPlusTree writeNodeToStore does not write full aligned block, causing Truncated node store IOException

Created by AI

## Description

`SoftBPlusTree.writeNodeToStore()` allocates a block-aligned region via `allocate()` (e.g., 64 bytes with `blockAlignment=64`) but only writes the actual serialized bytes (e.g., 21 bytes) to the backing store. The resulting `SoftNode` records `occupiedSpace` as the aligned size (64). When `SoftNode.resolve()` later reads from disk after GC clears the `SoftReference`, it tries to read `occupiedSpace` bytes — but the file only extends to `offset + serializedSize`, causing `FileChannel.read()` to return EOF.

This manifests as:

```
ISPN029019: Cannot load key k156901-5020813 from index
Caused by: java.io.IOException: Truncated node store at offset 134426
    at org.infinispan.persistence.sifs.Index$IndexFileNodeStore.read(Index.java:769)
    at org.infinispan.util.SoftBPlusTree$SoftNode.resolve(SoftBPlusTree.java:760)
```

## Why it's flaky

The gap between serialized data and aligned block end only matters for the **last block** written in `persistNewNodes()` — subsequent writes extend the file past the gap. The race window is between the last write to a segment's index file and the next `tree.put()` to that same segment. With 256 segments and 4 writer threads, this window is roughly every ~64 operations. GC must clear the `SoftReference` during that window, and a concurrent `cache.put()` → `index.getRecord()` → `tree.get()` must resolve that specific `SoftNode`.

Since `gcd(IndexEntry serialized size=12, blockAlignment=64) = 4` and the leaf header is 5 bytes (not divisible by 4), **every leaf node** has alignment padding, making the gap always present for the last node written.

## Reproducer

`SoftIndexFileStoreFunctionalTest.testWriteManyDifferentKeysAndIterate[256, true]([250000, true])` — fails intermittently in CI.

A deterministic unit test (`SoftBPlusTreeTest.testAlignmentPaddingWrittenToStore`) reproduces this using a `NodeStore` that simulates `FileChannel` EOF behavior (tracking the file high-water mark).

## Fix

In `writeNodeToStore()`, set `serializeBuffer.limit(space.occupiedSpace())` before writing so the full aligned block (including padding) is written to disk. The `serializeBuffer` capacity is also increased to `alignBlock(maxNodeSize)` to accommodate this.

--- a/core/src/main/java/org/infinispan/util/SoftBPlusTree.java
+++ b/core/src/main/java/org/infinispan/util/SoftBPlusTree.java
@@ -72,8 +72,10 @@ public class SoftBPlusTree<V> extends BPlusTree<V> {
       this.store = store;
       this.serializer = serializer;
       this.keyLoader = keyLoader;
-      this.serializeBuffer = ByteBuffer.allocate(maxNodeSize);
       this.blockAlignment = blockAlignment;
+      // Capacity must cover the aligned block size so writeNodeToStore can pad to occupiedSpace
+      int alignedMax = blockAlignment > 1 ? ((maxNodeSize + blockAlignment - 1) & -blockAlignment) : maxNodeSize;
+      this.serializeBuffer = ByteBuffer.allocate(alignedMax);
       this.initialStoreSize = initialStoreSize;
       this.storeSize = initialStoreSize;
    }
@@ -619,6 +621,10 @@ public class SoftBPlusTree<V> extends BPlusTree<V> {
       writeNode(node, serializer, serializeBuffer);
       serializeBuffer.flip();
       NodeSpace space = allocate((short) serializeBuffer.remaining());
+      // Write the full aligned block so the file extends to cover occupiedSpace.
+      // Without this, the file may be shorter than diskOffset + occupiedSpace when
+      // this is the last block written, causing EOF on SoftNode.resolve() after GC.
+      serializeBuffer.limit(space.occupiedSpace());
       store.write(serializeBuffer, space.offset);
       return space;
    }

--- a/core/src/test/java/org/infinispan/util/SoftBPlusTreeTest.java
+++ b/core/src/test/java/org/infinispan/util/SoftBPlusTreeTest.java
@@ -898,4 +898,84 @@ public class SoftBPlusTreeTest {
             ", after: " + sizeAfterReinsert,
             sizeAfterReinsert - sizeBeforeReinsert < sizeBeforeReinsert);
    }
+
+   // --- Alignment padding coverage test ---
+
+   /**
+    * A NodeStore that tracks the high-water mark of writes and throws IOException
+    * when reads go past the written extent, simulating FileChannel EOF behavior.
+    * This is how {@link org.infinispan.persistence.sifs.Index.IndexFileNodeStore}
+    * behaves: FileChannel.read() returns -1 when reading past the file's actual size.
+    */
+   static class FileSimulatingNodeStore implements SoftBPlusTree.NodeStore {
+      private byte[] data = new byte[4096];
+      private long fileSize = 0;
+
+      @Override
+      public void write(ByteBuffer buf, long offset) {
+         int length = buf.remaining();
+         ensureCapacity(offset + length);
+         buf.get(data, (int) offset, length);
+         long writeEnd = offset + length;
+         if (writeEnd > fileSize) {
+            fileSize = writeEnd;
+         }
+      }
+
+      @Override
+      public ByteBuffer read(long offset, int length) throws IOException {
+         if (offset >= fileSize) {
+            throw new IOException("Truncated node store at offset " + offset);
+         }
+         if (offset + length > fileSize) {
+            throw new IOException("Truncated node store at offset " + offset
+                  + " (requested " + length + " bytes, file size " + fileSize + ")");
+         }
+         byte[] result = new byte[length];
+         System.arraycopy(data, (int) offset, result, 0, length);
+         return ByteBuffer.wrap(result);
+      }
+
+      @Override
+      public void truncate(long size) {
+         fileSize = size;
+      }
+
+      private void ensureCapacity(long required) {
+         if (required > data.length) {
+            int newLen = Math.max((int) required, data.length * 2);
+            byte[] newData = new byte[newLen];
+            System.arraycopy(data, 0, newData, 0, data.length);
+            data = newData;
+         }
+      }
+   }
+
+   /**
+    * Reproduces the "Truncated node store at offset" bug. When a node's serialized
+    * size is not a multiple of the block alignment, writeNodeToStore() must write the
+    * full aligned block (not just the serialized bytes) so the file extends to cover
+    * the SoftNode's occupiedSpace. Without this fix, the last block written by
+    * persistNewNodes() leaves a gap between serialized data and the aligned block end,
+    * causing SoftNode.resolve() to hit EOF after GC clears the SoftReference.
+    */
+   public void testAlignmentPaddingWrittenToStore() throws IOException {
+      short blockAlignment = 64;
+      long headerSize = 26;
+      FileSimulatingNodeStore store = new FileSimulatingNodeStore();
+      SoftBPlusTree<Integer> tree = new SoftBPlusTree<>(MIN_NODE_SIZE, MAX_NODE_SIZE,
+            store, INT_SERIALIZER, keyLoader, blockAlignment, headerSize);
+
+      int count = 200;
+      for (int i = 0; i < count; i++) {
+         putTracked(tree, key(String.format("key-%05d", i)), i);
+      }
+
+      tree.clearSoftReferences();
+
+      for (int i = 0; i < count; i++) {
+         assertEquals("Value for key-" + String.format("%05d", i) + " should resolve from disk",
+               Integer.valueOf(i), tree.get(key(String.format("key-%05d", i))));
+      }
+   }
 }

